### ISSUE TITLE
[feature request] ai summerizer

### ISSUE THREAD

**Issue by celevra** [searxng/searxng#4097]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**
it would be great if we could integrate ollama or openai or whatever to summerize the search request with the first 3 search resultes returned by searxNG, like how brave search it does

![image](https://github.com/user-attachments/assets/f6862301-d982-459a-a3b8-5c40239b1e44)

**Comment by neon-mmd** [comment]
I think this issue is related to #2163. What do you say? :sweat_smile:

**Comment by LyndonArmitage** [comment]
If this is implemented it will definitely need to be an optional feature as plenty of people probably won't want to use an LLM due to cost, privacy or ideological reasons.

Implementation wise, from what I recall Searxng doesn't parse the whole HTML of a search result, and just presents a summary from the engine it is using. So would this AI summary be created using that limited information or use the full page contents? Alternatively, it could hand the page URL to the LLM to summarise, assuming it can crawl a web page.

**Comment by celevra** [comment]
i'm using open web ui to interact with ollama. In open web ui it is possible to activate websearch and send requests to searx, here is the wiki:
https://docs.openwebui.com/features/web_search/

that means it is allready possible (or code exists) to feed ollama with the help of the api from searx.
So we just need the way back into the searx frontend

**Comment by celevra** [comment]
if it helps i can provide access to a server running ollama, open-web-ui and searxNG

**Comment by tempname1024** [comment]
I'd written a quick implementation using the OpenAI schema (via OpenRouter), KaTeX, and marked to support Quick Answer functionality and replace my use of Kagi. OpenRouter requests are mediated by SearXNG and routed through a /quick_answer route to be streamed back to the client and rendered with search-supported inline citations and a references section. Users provide and store their own token via preferences. Quick Answer is triggered by ending a query in a "?".

If there's interest and this kind of feature isn't orthogonal to the goals of SearXNG, I'm happy to rework the plugin as needed to upstream the changes.

![2024-12-23_23-55](https://github.com/user-attachments/assets/db47b917-fc7d-4b00-833f-37df39ec1c35)

**Comment by celevra** [comment]
Yes that looks great! Is it possible to create a setting where the question mark isnt needed?

**Comment by tempname1024** [comment]
> Yes that looks great! Is it possible to create a setting where the question mark isnt needed?

Yes! An engine-variety !qa bang and toggle to run on every query would probably be more appropriate than the question mark route.

**Comment by return42** [comment]
> I'm happy to rework the plugin as needed to upstream the changes.

do you have a link to the sources?

**Comment by ankenu** [comment]
> > Yes that looks great! Is it possible to create a setting where the question mark isnt needed?
> 
> Yes! An engine-variety !qa bang and toggle to run on every query would probably be more appropriate than the question mark route.

Can you share with your plugin, please? 👀

**Comment by tempname1024** [comment]
> > > Yes that looks great! Is it possible to create a setting where the question mark isnt needed?
> > 
> > 
> > Yes! An engine-variety !qa bang and toggle to run on every query would probably be more appropriate than the question mark route.
> 
> Can you share with your plugin, please? 👀

Totally :) I'll work on creating a PR-ready branch and button down some loose ends in the implementation.

**Comment by barkoder** [comment]
If this feature is implemented please make it opt-in(with a url parameter or something). I am not okay with [the significant increase in water usage](https://arxiv.org/abs/2304.03271) involved in using "AI-powered" search.

Also I had a question regarding searches as they work currently in searxng.

When searxng is used to search, say google, does that query, in and of itself, trigger google's AI search? And searxng(currently) just chooses to omit serving that? So a higher water usage(on google's end) regardless of whether searxng has AI summary or not?

Or does using searxng without AI mean lower water use(on google's end)?

Thanks!

**Comment by tempname1024** [comment]
> If this feature is implemented please make it opt-in(with a url parameter or something). I am not okay with [the significant increase in water usage](https://arxiv.org/abs/2304.03271) involved in using "AI-powered" search.
> 
> Also I had a question regarding searches as they work currently in searxng.
> 
> When searxng is used to search, say google, does that query, in and of itself, trigger google's AI search? And searxng(currently) just chooses to omit serving that? So a higher water usage(on google's end) regardless of whether searxng has AI summary or not?
> 
> Or does using searxng without AI mean lower water use(on google's end)?
> 
> Thanks!

Agreed, the plugin should be disabled by default, and when enabled, invoked only when an AI summary is desired, rather than for every query.

![Image](https://github.com/user-attachments/assets/1afb7db4-62ec-49f4-b2d6-2e34b0f75475)

**Comment by celevra** [comment]
But what if i want a summary on every query?

**Comment by tempname1024** [comment]
The idea is since we don't rely on vector embeddings to determine whether a query's results can be augmented by an LLM (like Google, for example), and because there are many queries for which LLMs are unhelpful (and so needlessly expensive), deferring to the user's explicit invocation of the feature is probably the right way to go, and the route Mojeek, Kagi, et al. take as well.

**Comment by jeroenev** [comment]
Would probably implement it like this:
* Include a setting to enable/disable this feature (as is shown in the screenshot)
* include a setting to "Always show AI summary" (for users that always want the summary results)
* if "always show AI summary" is not enabled, it could be in a collapsable section (collapsed by default) that only loads the summary if the user opens the section

**Comment by barkoder** [comment]
@jeroenev If the AI summary is merely minimized, does that mean AI summary associated water is consumed on google's end, regardless of if a searxNG user chooses to hide or "always show" the AI summary?

**Comment by celevra** [comment]
@barkoder if it is disabled it is disabled....

**Comment by jeroenev** [comment]
@barkoder The idea is that the summary is only generated by AI when the user opens the collapsable section in that case, so nothing is sent to the AI until the user requests it by opening the section.
This to improve privacy / reduce API-cost / environmental impact (since no data is sent unless the user specifically wants it for that specific query)

**Comment by barkoder** [comment]
Thanks @jeroenev for the comprehensive reply. Does expanding that collapsible section and querying the AI summary require javascript to be enabled on the user's side?

**Comment by jeroenev** [comment]
I guess for the best experience it will need Javascript, so it can load the response in-page.
Could also be possible without, but without JS I think it has to reload the entire page to show content with the extra AI summary (extra url_param `show_ai_summary=1`)

**Comment by tempname1024** [comment]
Async proxied answers w/ requests built in the post_search context and reusing the infobox is probably the cleanest route; opaquely buffering responses server-side w/ additional UI components seems less user-friendly in my mind.

https://github.com/user-attachments/assets/5bc5641b-b9b0-4256-bcbb-9a6cc0092946

**Comment by mjtappert** [comment]
> i'm using open web ui to interact with ollama. In open web ui it is possible to activate websearch and send requests to searx, here is the wiki: https://docs.openwebui.com/features/web_search/
> 
> that means it is allready possible (or code exists) to feed ollama with the help of the api from searx. So we just need the way back into the searx frontend

I really, really like the idea of a hook between Ollama and SearXNG vs an online service for privacy. I know this would require a user to spin up an instance of Ollama, but that would be far preferable from my perspective. 

I do have one running, so if anyone has the skills that I do not, I'd be happy to help test / troubleshoot, etc

**Comment by tempname1024** [comment]
@mjtappert  You could try pointing the `quick_answer_api` config param in https://github.com/searxng/searxng/pull/4506 to Ollama's `http://localhost:11434/v1/chat/completions` endpoint, which should be [OpenAI-compatible](https://ollama.com/blog/openai-compatibility).

The rationale for providing first-class support for OpenRouter was to afford users of hosted instances the ability to choose which providers to trust, and to use whichever model(s) are best suited for their use case. The kind of [instructing following](https://github.com/searxng/searxng/pull/4506/commits/eaa96a0b6478baff605f3c7acfbc3072deb05db7#diff-c95404b89d4ec8acf144cbc9fcd6fe13d9808e79b2409f9fd4304b536e3042f6R26) we need for proper parsing and rendering of markdown, math, and references is pretty challenging for most models small enough to run on desktop GPUs.

**Comment by chandujr** [comment]
I've built a lightweight "proxy" that will add summaries of search results from SearXNG using OpenRouter. It works on top of your existing SearXNG installation, so if you are not using a heavily moded SearXNG, my proxy should work with it.

https://github.com/user-attachments/assets/108f9cf7-0690-4b6f-97fe-5032f3005a73

If anyone is interested I can upload the project to Github. It works with both SearXNG and 4get as of now.

**Comment by Trufax** [comment]
Iam very much interessted.

**Comment by chandujr** [comment]
> Iam very much interessted.

I have uploaded the repo. Please check if it works for you:
https://github.com/chandujr/search-results-summarizer

**Comment by cra88y** [comment]
@trufax check out my single file plugin for async rag summary :) compatible with multiple APIs https://github.com/cra88y/ai-answers-searxng

**Comment by Trufax** [comment]
@chandujr sorry for the delay but i directly jumped on the addon based solution from @cra88y which works like a charm. Thanks both of you for sharing.

**Comment by GregJohnStewart** [comment]
Bump to this! @chandujr 's solution looks great, but would love to see a more integrated / just searxng implementation, as this seems to wrap searxng completely?

**Comment by chandujr** [comment]
@GregJohnStewart Yes it does because I intended to make this compatible across multiple meta search engines. For your use case I think @cra88y 's [plugin](https://github.com/searxng/searxng/issues/4097#issuecomment-3757290162) would work better.
